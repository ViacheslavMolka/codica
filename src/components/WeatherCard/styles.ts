import { makeStyles } from '@mui/styles';

const useStyles = makeStyles(() => ({
  card: {
    maxWidth: 265,
    height: 370,
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
    margin: 10,
  },
  cardContent: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
}));

export default useStyles;
