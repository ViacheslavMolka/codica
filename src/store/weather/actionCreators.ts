import {
  GetWeatherAction,
  GetWeatherErrorAction,
  HourlyGetWeatherAction,
  UpdateGetWeatherAction,
  WeatherActionEnum
} from './actionTypes';

export const loadWeather = (payload: Array<object>): GetWeatherAction => ({ type: WeatherActionEnum.GET_WEATHER, payload });

export const newWeather = (payload: Array<object>): UpdateGetWeatherAction => ({ type: WeatherActionEnum.UPDATE_WEATHER, payload });

export const getHourlyWeather = (payload: Array<object>): HourlyGetWeatherAction => ({ type: WeatherActionEnum.HOURLY_WEATHER, payload });

export const getWeatherError = (payload: string): GetWeatherErrorAction => ({ type: WeatherActionEnum.ERROR, payload });
