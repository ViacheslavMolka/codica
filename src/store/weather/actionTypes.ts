export enum WeatherActionEnum {
  GET_WEATHER = "GET_WEATHER",
  UPDATE_WEATHER = "UPDATE_WEATHER",
  HOURLY_WEATHER = "HOURLY_WEATHER",
  ERROR = "ERROR",
}

export interface WeatherState {
  weather: Array<object>,
	hourly: Array<object>,
	error: string;
}

export interface GetWeatherAction {
  type: WeatherActionEnum.GET_WEATHER;
  payload: any;
}

export interface UpdateGetWeatherAction {
  type: WeatherActionEnum.UPDATE_WEATHER;
  payload: any;
}

export interface HourlyGetWeatherAction {
  type: WeatherActionEnum.HOURLY_WEATHER;
  payload: Array<object>;
}

export interface GetWeatherErrorAction {
  type: WeatherActionEnum.ERROR;
  payload: string;
}

export type WeatherActions =
  GetWeatherAction |
  UpdateGetWeatherAction |
  HourlyGetWeatherAction |
  GetWeatherErrorAction
