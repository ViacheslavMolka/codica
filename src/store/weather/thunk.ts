import { AppDispatch } from "..";
import {
  loadWeather,
  newWeather,
  getHourlyWeather,
  getWeatherError,
} from "./actionCreators";

export const getWeather = (name: string) => {
	return async (dispatch: AppDispatch) => {
    try {
      const res = await fetch(`https://api.openweathermap.org/data/2.5/weather?q=${name}&appid=76219c903b8e8758872f9c9ead42c0de`);
      const data = await res.json();
      if (data.message) {
        dispatch(getWeatherError(data.message))
      } else {
        dispatch(loadWeather(data))
      }
    } catch(e) {
      dispatch(getWeatherError("Error"))
    }
	};
};

export const updateWeather = (name: string) => {
	return async (dispatch: AppDispatch) => {
    try {
      const res = await fetch(`https://api.openweathermap.org/data/2.5/weather?q=${name}&appid=76219c903b8e8758872f9c9ead42c0de`);
      const data = await res.json();
      if (data.message) {
        dispatch(getWeatherError(data.message))
      } else {
        dispatch(newWeather(data))
      }
    } catch(e) {
      dispatch(getWeatherError("Error"))
    }
	};
};

export const hourlyWeather = (name: string) => {
	return async (dispatch: AppDispatch) => {
    try {
      const res = await fetch(`https://api.openweathermap.org/data/2.5/forecast?q=${name}&appid=76219c903b8e8758872f9c9ead42c0de`);
      const data = await res.json();
      if (data.message) {
        dispatch(getWeatherError(data.message))
      } else {
        dispatch(getHourlyWeather(data))
      }
    } catch(e) {
      dispatch(getWeatherError("Error"))
    }
	};
};
