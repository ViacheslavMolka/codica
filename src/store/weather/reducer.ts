/* eslint-disable @typescript-eslint/no-explicit-any */
import { WeatherActionEnum, WeatherState, WeatherActions, IWeather } from './actionTypes';

const initialState: WeatherState = {
	weather: [],
	hourly: [],
	error: '',
};

export default function weather(state = initialState, action: WeatherActions): WeatherState {
	switch (action.type) {
		case WeatherActionEnum.GET_WEATHER: {
			const isHas: IWeather | any = state.weather.find((item: any) => item.name === action.payload.name);
			const checkWeather = isHas ? [] : [action.payload];
      return {
				...state,
        weather: [...state.weather, ...checkWeather],
      }}
		case WeatherActionEnum.UPDATE_WEATHER: {
			const newWeather: Array<IWeather> = state.weather.map((item: any) => {
				if (item.name === action.payload.name) {
					return action.payload
				}
				return item;
			})
			return {
				...state,
				weather: newWeather,
			}
		}
		case WeatherActionEnum.HOURLY_WEATHER:
      return {
				...state,
        hourly: [...state.hourly, action.payload],
      }
		case WeatherActionEnum.ERROR:
			return {
				...state,
				error: action.payload,
			}
		default:
			return state;
	}
}
