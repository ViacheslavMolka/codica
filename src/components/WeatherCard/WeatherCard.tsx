import React from 'react';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import CardMedia from '@mui/material/CardMedia';
import Typography from '@mui/material/Typography';
import { Button, CardActionArea, CardActions } from '@mui/material';

import WeatherBar from '../WeatherBar/WeatherBar';
import Weather from './weather.jpg';
import { toСelsius } from '../../helpers';
import { useAppDispatch } from '../../store';
import { hourlyWeather, updateWeather } from '../../store/weather/thunk';
import { useTypedSelector } from '../../hooks/useTypedSelector';
import useStyles from './styles';

const WeatherCard: React.FC<any> = ({ info }) => {
  const [isDetails, setIsDetails] = React.useState(true);
  const dispatch = useAppDispatch();
  const classes = useStyles();
  const { hourly = [] } = useTypedSelector(state => state.weather);

  React.useEffect(() => {
    dispatch(hourlyWeather(info.name));
  }, [info.name]);

  const temp: any = hourly.length && hourly.find((item: any) => item?.city?.name === info?.name);
  
  return (
    <div>
      {info && (isDetails ? 
        <Card className={classes.card}>
          <CardActionArea onClick={() => setIsDetails(!isDetails)}>
            <CardMedia
              component="img"
              image={Weather}
            />
            <CardContent className={classes.cardContent}>
              <Typography gutterBottom variant="h5" component="div">
                {info?.name}
              </Typography>
              <Typography variant="h5" color="GrayText">
                {toСelsius(info?.main?.temp)}°
              </Typography>
              <Typography variant="subtitle2" color="GrayText">
                {info?.clouds?.all > 50 ? 'Cloudy' : 'Cloudless'}
              </Typography>
            </CardContent>
          </CardActionArea>
          <CardActions>
            <Button
              sx={{ margin: '0 auto' }}
              size="small"
              color="inherit"
              onClick={() => dispatch(updateWeather(info.name))}
            >
              Update
            </Button>
          </CardActions>
        </Card>
          :
        <Card className={classes.card}>
          <CardActionArea onClick={() => setIsDetails(!isDetails)}>
            <CardContent>
              <Typography gutterBottom variant="h5" component="div">
                Details
              </Typography>
              <Typography sx={{ paddingTop: 3 }} variant="body2" color="text.secondary">
                Feels like: {toСelsius(info?.main?.feels_like)}°
              </Typography>
              <Typography variant="body2" color="text.secondary">
                Humidity: {info?.main?.humidity}%
              </Typography>
              <Typography variant="body2" color="text.secondary">
                Wind speed: {info?.wind?.speed} m/s
              </Typography>
              <Typography variant="body2" color="text.secondary">
                Sunrise: {new Date(info?.sys?.sunrise).toString().substr(15, 9)} AM
              </Typography>
              <Typography variant="body2" color="text.secondary">
                Sunset: {new Date(info?.sys?.sunset).toString().substr(15, 9)} PM
              </Typography>
            </CardContent>
              <WeatherBar temp={temp?.list ? temp.list.slice(0, 7) : []} />
          </CardActionArea>
        </Card>
      )}
    </div>
  );
};

export default WeatherCard;
