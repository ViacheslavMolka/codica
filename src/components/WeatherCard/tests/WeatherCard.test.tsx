import * as React from 'react';
import * as reactRedux from 'react-redux';
import renderer from 'react-test-renderer';
import { combineReducers, createStore } from 'redux';
import { Provider } from 'react-redux';

import weather from '../../../store/weather/reducer';
import WeatherCard from '../WeatherCard';

const mockedProps = {
  info: {
    name: 'Kyiv',
    main: {
      temp: '283',
      feels_like: '287',
      humidity: 60,
  },
    wind: {
      speed: 7.7
  },
    sys: {
      sunset: 123123131,
      sunrise: 123123131,
    }
  }
};

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('components > WeatherCard component', () => {
  const useSelectorMock = reactRedux.useSelector;
  const store = createStore(combineReducers({ weather }));
  const useTypedSelector: any = useSelectorMock;
  const mockStore = {
    hourly: [{ name: 'Lviv' }],
    weather: [],
    error: '',
  };
  beforeEach(() => {
    useTypedSelector.mockImplementation((selector: any) => selector(mockStore));
  })
  afterEach(() => {
    useTypedSelector.mockClear();
  })

  it('should render WeatherCard component', () => {
    const renderedComponnet = renderer.create(
      <Provider store={store}>
        <WeatherCard {...mockedProps}/>
      </Provider>
    );
    expect(renderedComponnet).toMatchSnapshot('should render correctly');
  });
});
