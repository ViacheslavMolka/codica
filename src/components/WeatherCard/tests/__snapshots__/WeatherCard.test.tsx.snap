// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components > WeatherCard component should render WeatherCard component: should render correctly 1`] = `
<div>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root makeStyles-card-1 css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <button
      className="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <img
        className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
        src="weather.jpg"
      />
      <div
        className="MuiCardContent-root makeStyles-cardContent-2 css-46bh2p-MuiCardContent-root"
      >
        <div
          className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
        >
          Kyiv
        </div>
        <h5
          className="MuiTypography-root MuiTypography-h5 css-1dlrolv-MuiTypography-root"
        >
          10
          Â°
        </h5>
        <h6
          className="MuiTypography-root MuiTypography-subtitle2 css-10i9if-MuiTypography-root"
        >
          Cloudless
        </h6>
      </div>
      <span
        className="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
      />
    </button>
    <div
      className="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
    >
      <button
        className="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorInherit MuiButtonBase-root  css-g0zxa1-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Update
      </button>
    </div>
  </div>
</div>
`;
